### ISSUE TITLE
Add Anthropic integration

### ISSUE THREAD

**Issue by davidmigloz** [davidmigloz/langchain_dart#316]
- [ChatAnthropic](https://python.langchain.com/docs/integrations/chat/anthropic)

**Comment by cyberpwnn** [comment]
Literally no idea why this took so long or why the package even exists but hasnt been touched in forever but i slapped this together... 

Note: Anthropic doesnt have their cors policy setup to allow any origin so if your using this for flutter directly calling it it wont work when running flutter on web. Other than that works like a charm, streams just fine. 

Note: Using tiktoken encoding `cl100k_base` because anthropic doesnt have a public tokenizer... or i couldnt find something usable for it. 

Just put this anywhere in your project and use it like any other langchain extension. 

```dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:langchain/langchain.dart';
import 'package:langchain_tiktoken/langchain_tiktoken.dart';
import 'package:uuid/uuid.dart';

class ChatAnthropicOptions extends ChatModelOptions {
  /// See https://docs.anthropic.com/claude/docs/models-overview
  final String model;
  final int maxTokens; // 1024?
  final String? user; // inside metadata
  final List<String>? stopSequence;
  final double? temperature; // 0 to 1, defaults to 1.0
  final int? topK;
  final int? topP;

  const ChatAnthropicOptions(
      {this.model = 'claude-3-haiku-20240307',
      this.maxTokens = 1024,
      this.user,
      this.stopSequence,
      this.temperature,
      this.topK,
      this.topP});

  Map<String, dynamic> _decorateRequest(
          {required Map<String, dynamic> request,
          bool? stream,
          String? systemPrompt}) =>
      {
        ...request,
        'model': model,
        'max_tokens': maxTokens,
        if (user != null) 'metadata': {'user_id': user},
        if (stopSequence != null) 'stop_sequence': stopSequence,
        if (stream != null) 'stream': stream,
        if (systemPrompt != null) 'system': systemPrompt,
        if (temperature != null) 'temperature': temperature,
        if (topK != null) 'top_k': topK,
        if (topP != null) 'top_p': topP,
      };
}

class ChatAnthropic extends BaseChatModel<ChatAnthropicOptions> {
  final String apiKey;
  final http.Client? client;
  late http.Client _client;

  ChatAnthropic({
    required this.apiKey,
    this.client,
    super.defaultOptions = const ChatAnthropicOptions(),
  }) {
    _client = client ?? http.Client();
  }

  @override
  String get modelType => 'chat-anthropic';

  Map<String, dynamic> _toAnthropicMessage(ChatMessage message) => {
        'role': message is AIChatMessage ? 'assistant' : 'user',
        'content': message.contentAsString,
      };

  @override
  Stream<ChatResult> stream(
    PromptValue input, {
    ChatAnthropicOptions? options,
  }) async* {
    options ??= defaultOptions;
    http.StreamedResponse response = await _client.send(
        http.Request('POST', Uri.parse('https://api.anthropic.com/v1/messages'))
          ..headers.addAll({
            'Access-Control-Allow-Origin': '*',
            'x-api-key': apiKey,
            'anthropic-version': '2023-06-01',
            'content-type': 'application/json'
          })
          ..body = jsonEncode(options._decorateRequest(
              systemPrompt: _toSystemMessage(input, options: options),
              stream: true,
              request: {'messages': _toMessages(input, options: options)})));

    Stream<String> lines = response.stream
        .transform(const Utf8Decoder())
        .transform(const LineSplitter());

    String event = "none";
    int inputTokens = 0;
    int outputTokens = 0;
    await for (final line in lines) {
      if (line.startsWith("event: ")) {
        event = line.split(":")[1].trim();
      } else if (line.startsWith("data: ") && event == "content_block_delta") {
        final Map<String, dynamic> data = jsonDecode(line.substring(6));
        yield ChatResult(
            id: const Uuid().v4(),
            output: AIChatMessage(content: data["delta"]["text"]),
            finishReason: FinishReason.unspecified,
            metadata: data,
            streaming: true,
            usage: LanguageModelUsage(
              promptTokens: inputTokens,
              responseTokens: outputTokens,
            ));
        inputTokens = 0;
        outputTokens = 0;
      } else if (line.startsWith("data: ") && event == "message_delta") {
        final Map<String, dynamic> data = jsonDecode(line.substring(6));
        Map<String, dynamic> usage = data["usage"] ?? {};
        inputTokens += (usage["input_tokens"] as int?) ?? 0;
        outputTokens += (usage["output_tokens"] as int?) ?? 0;
      }

      yield ChatResult(
          id: "done",
          output: const AIChatMessage(content: ""),
          finishReason: FinishReason.unspecified,
          metadata: const {},
          streaming: true,
          usage: LanguageModelUsage(
            promptTokens: inputTokens,
            responseTokens: outputTokens,
          ));
    }
  }

  List<Map<String, dynamic>> _toMessages(PromptValue input,
      {ChatAnthropicOptions? options}) {
    options ??= defaultOptions;
    List<Map<String, dynamic>> messages;

    if (input is ChatPromptValue) {
      messages = input.messages
          .where((m) => m is AIChatMessage || m is HumanChatMessage)
          .where((m) => m.contentAsString.trim().isNotEmpty)
          .map(_toAnthropicMessage)
          .toList();
    } else {
      messages = [
        _toAnthropicMessage(HumanChatMessage(
            content: ChatMessageContentText(text: input.toString())))
      ];
    }

    return messages;
  }

  String? _toSystemMessage(PromptValue input, {ChatAnthropicOptions? options}) {
    if (input is ChatPromptValue) {
      Iterable<SystemChatMessage> systemMessages =
          input.messages.whereType<SystemChatMessage>();

      if (systemMessages.isNotEmpty) {
        return input.messages
            .whereType<SystemChatMessage>()
            .map((s) => s.content)
            .join('\n');
      }
    }

    return null;
  }

  @override
  Future<ChatResult> invoke(PromptValue input,
      {ChatAnthropicOptions? options}) async {
    options ??= defaultOptions;
    http.Response response = await _client.post(
        (Uri.parse('https://api.anthropic.com/v1/messages')),
        headers: {
          'Access-Control-Allow-Origin': '*',
          'x-api-key': apiKey,
          'anthropic-version': '2023-06-01',
          'content-type': 'application/json'
        },
        body: jsonEncode(options._decorateRequest(
            systemPrompt: _toSystemMessage(input, options: options),
            stream: false,
            request: {'messages': _toMessages(input, options: options)})));

    if (response.statusCode != 200) {
      throw Exception('Failed to invoke anthropic: ${response.body}');
    }

    final Map<String, dynamic> data = jsonDecode(response.body);

    return ChatResult(
        id: data["id"],
        output: AIChatMessage(
            content: (data["content"] as List<dynamic>).last["text"]),
        finishReason: FinishReason.unspecified,
        metadata: data,
        usage: LanguageModelUsage(
          promptTokens: data["usage"]["input_tokens"],
          responseTokens: data["usage"]["output_tokens"],
        ));
  }

  @override
  Future<List<int>> tokenize(PromptValue promptValue,
      {ChatAnthropicOptions? options}) async {
    // using tiktoken for now
    final encoding = getEncoding("cl100k_base");
    return encoding.encode(promptValue.toString());
  }
}
```

**Comment by davidmigloz** [comment]
Hey @cyberpwnn,

Thanks for the code and suggestions!

I created the issue months ago, hoping someone else would pick it up. I didn't have the bandwidth to work on it myself (and I didn't need it, as I was using other models).

But luckily, @alfredobs97 is working on it at the moment. We've implemented a client (https://pub.dev/packages/anthropic_sdk_dart) and will integrate it into LangChain.dart in the next release.
